### PR TITLE
44 dbt core 13 upgrade incremental mats more standard and more error proof

### PR DESCRIPTION
**Pull request by ArgusLi**
### Summary

Upgrade incremental materialization to new dbt-core 1.3 logic while retaining Dremio specific changes.

### Description

- Add valid_incremental_strategies() with append being the only valid strategy.
- Refactor materialization to match new default incremental materialization.
- Add macro dremio__rename_relation() that uses CTAS and DROP instead of default ALTER TABLE and RENAME to.

### Test Results

30/30 tests passed.

### Changelog

-   [x] Added a summary of what this PR accomplishes to CHANGELOG.md

### Related Issue
[44](https://github.com/dremio/dbt-dremio/issues/44)

### CODE REVIEW

File: dbt/include/dremio/macros/materializations/incremental/incremental.sql [hunk 1]
```diff
@@ -14,64 +14,96 @@ limitations under the License.*/
...
+    
   {% endif %}
 
   {%- call statement('main') -%}
     {{ build_sql }}
   {%- endcall -%}
 
-  {% if not(old_relation is none or full_refresh_mode) %}
-    {{ adapter.drop_relation(target_tmp_relation) }}
+  {% if need_swap %}
```
Comment (ravjotbrar) [inline dbt/include/dremio/macros/materializations/incremental/incremental.sql:78]: I noticed rename_relation is being used in view and table macros as well. Is this something we should be doing as well?

Comment (ravjotbrar) [reply dbt/include/dremio/macros/materializations/incremental/incremental.sql:78]: This might be a separate ticket, but it'll keep our macros more up to date.

File: dbt/include/dremio/macros/materializations/incremental/incremental.sql [hunk 2]
```diff
@@ -14,64 +14,96 @@ limitations under the License.*/
...
+    {% set build_sql = get_create_table_as_sql(False, target_relation, sql) %}
   {% elif full_refresh_mode %}
-    {% do adapter.drop_relation(old_relation) %}
-    {% set build_sql = create_table_as(False, target_relation, external_query(sql)) %}
+    {% set build_sql = get_create_table_as_sql(False, intermediate_relation, sql) %}
+    {% set need_swap = true %}
   {% else %}
-    {% if tmp_relation is not none %}
-      {{ adapter.drop_relation(tmp_relation) }}
+    {% do run_query(get_create_table_as_sql(True, temp_relation, sql)) %}
```
Comment (ravjotbrar) [inline dbt/include/dremio/macros/materializations/incremental/incremental.sql:54]: the old run_query had the sql wrapped in an external_query macro. Is there a reason we removed this?

File: dbt/include/dremio/macros/materializations/incremental/incremental.sql [hunk 3]
```diff
@@ -14,64 +14,96 @@ limitations under the License.*/
...
+  {{ run_hooks(pre_hooks, inside_transaction=False) }}
+ 
+  -- `BEGIN` happens here:
+  {{ run_hooks(pre_hooks, inside_transaction=True) }}
 
-  {{ run_hooks(pre_hooks) }}
+  {% set to_drop = [] %}
 
-  {% if old_relation is none %}
-    {% set build_sql = create_table_as(False, target_relation, external_query(sql)) %}
```
Comment (ravjotbrar) [inline dbt/include/dremio/macros/materializations/incremental/incremental.sql:45]: The old build_sql  had the sql wrapped in an external_query macro. Is there a reason we removed this?

File: dbt/include/dremio/macros/materializations/incremental/incremental.sql [hunk 4]
```diff
@@ -14,64 +14,96 @@ limitations under the License.*/
...
 {% materialization incremental, adapter='dremio' -%}
 
-  {%- set identifier = model['alias'] -%}
-  {%- set old_relation = adapter.get_relation(database=database, schema=schema, identifier=identifier) -%}
-  {%- set target_relation = api.Relation.create(identifier=identifier,
-                                                schema=schema,
-                                                database=database,
-                                                type='table') -%}
-
-  {%- set tmp_identifier = model['name'] + '__dbt_tmp' -%}
```
Comment (ravjotbrar) [inline dbt/include/dremio/macros/materializations/incremental/incremental.sql:24]: Is there a reason we're getting rid of the tmp_identifier?

File: dbt/include/dremio/macros/materializations/incremental/incremental.sql [hunk 5]
```diff
@@ -14,64 +14,96 @@ limitations under the License.*/
...
-                                                database=database,
-                                                type='table') -%}
-
-  {%- set raw_strategy = config.get('incremental_strategy', validator=validation.any[basestring]) or 'append' -%}
-  {%- set raw_file_format = config.get('format', validator=validation.any[basestring]) or 'iceberg' -%}
-  {%- set file_format = dbt_dremio_validate_get_file_format(raw_file_format) -%}
-  {%- set strategy = dbt_dremio_validate_get_incremental_strategy(raw_strategy, file_format) -%}
+  -- relations
+  {%- set existing_relation = load_cached_relation(this) -%} 
+  {%- set target_relation = this.incorporate(type='table') -%}
```
Comment (ravjotbrar) [inline dbt/include/dremio/macros/materializations/incremental/incremental.sql:19]: We should be careful about using dbt's defaults for these macros. In the end, we as a team need to add more dremio specific test cases around custom schemas and different targets, to ensure changes like these don't break any of the edge cases.

File: dbt/include/dremio/macros/materializations/incremental/incremental.sql (hunk 1 above)
Comment (ArgusLi) [reply dbt/include/dremio/macros/materializations/incremental/incremental.sql:78]: rename_relation is used here for full_refresh_mode to fully replace the relation. I don't see a real need to use it in other macros unless its functionality is needed.

File: dbt/include/dremio/macros/materializations/incremental/incremental.sql (hunk 4 above)
Comment (ArgusLi) [reply dbt/include/dremio/macros/materializations/incremental/incremental.sql:24]: tmp_identifier was used in the old logic to get tmp_relation and create target_tmp_relation. The new logic has moved to temp_relation and intermediate_relation with a different relation creation strategy, therefore, it is no longer needed.

File: dbt/include/dremio/macros/adapters/relation.sql [hunk 6]
```diff
@@ -25,3 +25,12 @@ limitations under the License.*/
     drop {{ relation.type }} if exists {{ relation }}
   {%- endcall %}
 {% endmacro %}
+
+{% macro dremio__rename_relation(from_relation, to_relation) -%}
+  {% call statement('rename_relation1/2 - create to_relation from from_relation') -%}
+    CREATE TABLE {{ to_relation }} AS (select * from {{from_relation}})
+  {%- endcall %}
+  {% call statement('rename_relation2/2 - drop from_relation') -%}
```
[the remark this answers is not in the document]
Comment (ArgusLi) [reply dbt/include/dremio/macros/adapters/relation.sql:33]: I didn't think of that. Good catch. However, I do not like the dremio__create_table_as macro, there is an extra parameter "temporary" that does not get used at all in the macro. I will give both a try.

File: dbt/include/dremio/macros/materializations/incremental/incremental.sql (hunk 5 above)
Comment (ArgusLi) [reply dbt/include/dremio/macros/materializations/incremental/incremental.sql:19]: On the contrary, I think incorporating new dbt logic is necessary. The new dbt logic is more elegant and less "hacky". It's also not like we're just using dbt's default, we're using its logic but keeping Dremio specific changes. In the future I think the ideal direction is actually using dbt's defaults for everything and just having Dremio specific macros. That way we do not have to maintain any of these materialization strategies - that will be done by the dbt team.

However, I agree we should have more Dremio specific test cases.

File: dbt/include/dremio/macros/materializations/incremental/incremental.sql (hunk 3 above)
Comment (ArgusLi) [reply dbt/include/dremio/macros/materializations/incremental/incremental.sql:45]: We really should be using external_query. I will change this. Thanks!

File: dbt/include/dremio/macros/materializations/incremental/incremental.sql (hunk 2 above)
Comment (ArgusLi) [reply dbt/include/dremio/macros/materializations/incremental/incremental.sql:54]: We really should be using external_query. I will change this. Thanks!

File: dbt/include/dremio/macros/adapters/relation.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (ArgusLi) [reply dbt/include/dremio/macros/adapters/relation.sql:33]: dremio__create_table() has been implemented, but dremio__drop_table() leads to an error and will be addressed in https://github.com/dremio/dbt-dremio/issues/113.

**Review by ravjotbrar — APPROVED** [review]
